src/NLog/Internal/PropertiesDictionary.cs

```diff
@@ -0,0 +1,698 @@
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE 
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF 
+// THE POSSIBILITY OF SUCH DAMAGE.
+// 
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+
+namespace NLog.Internal
+{
+    /// <summary>
+    /// Combines the MessageTemplate-Properties with LogEvent-Properties into a single dictionary
+    ///  - MessageTemplate-Properties are always returned first in the order found in the MessageTemplate
+    ///  - LogEvent-Properties has higher priority than the MessageTemplate-Properties
+    /// </summary>
+    internal sealed class PropertiesDictionary : IDictionary<object, object>, IDictionary
+    {
+        private IList<KeyValuePair<string,object>> _messageProperties;
+        private Dictionary<object, object> _eventProperties;
+
+        private Dictionary<object, object> EventProperties { get { return _eventProperties ?? (_eventProperties = new Dictionary<object, object>()); } }
+
+        public IList<KeyValuePair<string, object>> MessageProperties
+        {
+            get { return _messageProperties; }
+            set
+            {
+                if (value != null && value.Count > 0)
+                {
+                    if (_eventProperties != null && _eventProperties.Count > 0)
+                    {
+                        for (int i = 0; i < value.Count; ++i)
+                        {
+                            _eventProperties.Remove(value[i].Key);
+                        }
+                    }
+
+                    _messageProperties = MakePropertiesUnique(value);
+                }
+                else
+                {
+                    _messageProperties = value ?? ArrayHelper.Empty<KeyValuePair<string, object>>();
+                }
+            }
+        }
+
+        private int MessagePropertiesIndexOfKey(object key)
+        {
+            if (_messageProperties != null)
+            {
+                string keyString = key as string;
+                if (keyString != null)
+                {
+                    for (int i = 0; i < _messageProperties.Count; ++i)
+                    {
+                        if (string.Equals(_messageProperties[i].Key, keyString, StringComparison.Ordinal))
+                        {
+                            return i;
+                        }
+                    }
+                }
+            }
+            return -1;
+        }
+
+        private static IList<KeyValuePair<string, object>> MakePropertiesUnique(IList<KeyValuePair<string, object>> value)
```

Comment:
please private methods below the publics

Reply:
Fixed